plugins/repository-gcs/qa/google-cloud-storage/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageTestServer.java

```diff
@@ -224,33 +229,150 @@ public Response handle(final String method,
+            final String uploadType = params.get("uploadType");
+            if ("resumable".equals(uploadType)) {
+                final String objectName = params.get("name");
+                if (Strings.hasText(objectName) == false) {
+                    return newError(RestStatus.INTERNAL_SERVER_ERROR, "object name is missing");
+                }
+                final Bucket bucket = buckets.get(params.get("bucket"));
+                if (bucket == null) {
+                    return newError(RestStatus.NOT_FOUND, "bucket not found");
+                }
+                if (bucket.objects.put(objectName, EMPTY_BYTE) == null) {
+                    final String location = endpoint + "/upload/storage/v1/b/" + bucket.name + "/o?uploadType=resumable&upload_id="
+                            + objectName;
+                    return new Response(RestStatus.CREATED, singletonMap("Location", location), XContentType.JSON.mediaType(), EMPTY_BYTE);
+                } else {
+                    return newError(RestStatus.CONFLICT, "object already exist");
+                }
+            } else if ("multipart".equals(uploadType)) {
+/* A multipart/related request body looks like this (note the binary dump inside a text blob! nice!):
+*--__END_OF_PART__
+*Content-Length: 135
+*Content-Type: application/json; charset=UTF-8
+*content-transfer-encoding: binary
+*
+*{"bucket":"bucket_test","crc32c":"7XacHQ==","md5Hash":"fVztGkklMlUamsSmJK7W+w==","name":"tests-KEwE3bU4TuyetBgQIghmUw/master.dat-temp"}
+*--__END_OF_PART__
+*content-transfer-encoding: binary
+*
+*KEwE3bU4TuyetBgQIghmUw
+*--__END_OF_PART__--
+*/
+                String boundary = "__END_OF_PART__";
+                // Determine the multipart boundary
+                final List<String> contentTypes = headers.getOrDefault("Content-Type", headers.get("Content-type"));
+                if (contentTypes != null) {
+                    final String contentType = contentTypes.get(0);
+                    if ((contentType != null) && contentType.contains("multipart/related; boundary=")) {
+                        boundary = contentType.replace("multipart/related; boundary=", "");
+                    }
+                }
+                InputStream inputStreamBody = new ByteArrayInputStream(body);
+                final List<String> contentEncodings = headers.getOrDefault("Content-Encoding", headers.get("Content-encoding"));
+                if (contentEncodings != null) {
+                    if (contentEncodings.stream().anyMatch(x -> "gzip".equalsIgnoreCase(x))) {
+                        inputStreamBody = new GZIPInputStream(inputStreamBody);
+                    }
+                }
+                // Read line by line ?both? parts of the multipart
+                try (BufferedReader reader = new BufferedReader(new InputStreamReader(inputStreamBody, StandardCharsets.ISO_8859_1))) {
```

Comment:
Mixing UTF-8 encoded text (variable length encoding) with binary data in the same entity is a sure source of bugs and headaches 💯 
Decoding text as ISO_8859_1 simplifies the parsing.